### ISSUE TITLE
Bil Store Eval behaves badly

### ISSUE THREAD

**Issue by DIJamner** [BinaryAnalysisPlatform/bap#848]
In the following code, `Exp.eval read` should return `0x1:16u`. However, it returns `0x101:16u`. Note that this behavior is only observable on pull request #847 since without that fix, it will give a type error at runtime.
```
open Bap.Std;;
open Bil;;
let a1 = int (Word.zero 64);;
let a2 = int (Word.one 64);;
let m = (unknown "mem" (Type.mem `r64 `r16));;
let m1 = store ~mem:m ~addr:a1 (int (Word.one 16)) LittleEndian `r16;;
let m2 = store ~mem:m1 ~addr:a2 (int (Word.one 16)) LittleEndian `r16;;
let read = load ~mem:m2 ~addr:a1 LittleEndian `r32;;
Type.infer read;;
Exp.eval read;;
```
This is likely due to incorrect behavior of `store` on memories with a cell size other than 8 bits.

**Comment by ivg** [comment]
I believe you meant `r16 in the last read, as `r32 would result in the undefined value, since we have only three defined bytes

**Comment by ivg** [comment]
While I believe that this issue should be fixed in BAP 2.0 in a sense that the semantics of loads and stores are now correct in the Core Theory and Core Theory terms should be reified into BIL loads and stores correctly (and vice versa), but it looks like that `Exp.eval` is still broken, but differently. Now, `Exp.eval read` doesn't terminate.
